### PR TITLE
feat: add a new API endpoint to bind all unbound workspaces to autocreated namespaces

### PR DESCRIPTION
**Pull request by salonig23**
## Ticket

DET-10226

## Checklist

- [ ] Changes have been manually QA'd
- [ ] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: master/internal/api_workspace.go
Comment (corban-beaird) [inline master/internal/api_workspace.go:864]: is this still intended to be commented out?

Comment (maxrussell) [reply master/internal/api_workspace.go:864]: I think she's waiting on my updates for the licensing.

**Review by maxrussell — CHANGES_REQUESTED** [review]
Still reviewing, but posting this as-is for now to get conversation started

**Comment by corban-beaird** [comment]
Just spit balling, but would it be worth having these structured as instead 2 endpoints, a GET & a POST, where the get will return the set of workspace_ids that aren't associated with a namespace & then a POST responsible for the auto-creation &  binding of the namespace to the given workspaces based on the provided ids? The POST could potentially be framed as a bulk action endpoint. 

My concern is this seems like a very niche endpoint that will likely only get used once per determined instance. It's absolutely okay to disagree, overall the code looks solid!

**Comment by maxrussell** [comment]
I agree and had the same thought, Corban. I prefer explicit APIs over implicit ones. And in my experience it's usually worth the performance hit of roundtripping the list of ids back to a client. Something like
```proto
  // Binds all unbound workspaces to new auto-created namespaces.
  rpc GetUnboundWorkspaces(
      GetUnboundWorkspacesRequest)
      returns (GetUnboundWorkspacesResponse) {
    option (google.api.http) = {
      get: "/api/v1/workspaces/unbound",
    };
    option (grpc.gateway.protoc_gen_swagger.options.openapiv2_operation) = {
      tags: "Workspaces"
    };
  }

  // Binds the specified workspaces to new auto-created namespaces if they're unbound.
  rpc BulkAutoCreateWorkspaceNamespaceBindings(
      BulkAutoCreateWorkspaceNamespaceBindingsRequest)
      returns (BulkAutoCreateWorkspaceNamespaceBindingsResponse) {
    option (google.api.http) = {
      post: "/api/v1/workspaces/auto-namespace-bindings",
    };
    option (grpc.gateway.protoc_gen_swagger.options.openapiv2_operation) = {
      tags: "Workspaces"
    };
  }

// Request for finding all unbound workspaces.
message GetUnboundWorkspacesRequest {}

// Response to GetUnboundWorkspacesRequest.
message GetUnboundWorkspacesResponse {
  // The list of returned workspace ids.
  repeated int32 = 1;
  // Pagination information of the full dataset.
  Pagination pagination = 2;
}

// Request for binding the specified unbound workspaces to new auto-created namespaces.
message BulkAutoCreateWorkspaceNamespaceBindingsRequest {
  // The list of returned workspace ids.
  repeated int32 = 1;
}

// Response to BulkAutoCreateWorkspaceNamespaceBindingsRequest.
message BulkAutoCreateWorkspaceNamespaceBindingsResponse {}
```

**Comment by salonig23** [comment]
> Just spit balling, but would it be worth having these structured as instead 2 endpoints, a GET & a POST, where the get will return the set of workspace_ids that aren't associated with a namespace & then a POST responsible for the auto-creation & binding of the namespace to the given workspaces based on the provided ids? The POST could potentially be framed as a bulk action endpoint.
> 
> My concern is this seems like a very niche endpoint that will likely only get used once per determined instance. It's absolutely okay to disagree, overall the code looks solid!

That's a good point @corban-beaird @maxrussell . The purpose of this API endpoint is to be just used once when the Client is migrating to this new concept of workspace namespace bindings. The assumption is that all their workspaces will be unbound at the time. Does that change your thinking?

**Comment by maxrussell** [comment]
@salonig23: it doesn't change my thinking. Even thousands of workspace ids in a request/response body wouldn't be a particularly big request in my opinion.

**Comment by salonig23** [comment]
> I agree and had the same thought, Corban. I prefer explicit APIs over implicit ones. And in my experience it's usually worth the performance hit of roundtripping the list of ids back to a client. Something like
> 
> ```proto
>   // Binds all unbound workspaces to new auto-created namespaces.
>   rpc GetUnboundWorkspaces(
>       GetUnboundWorkspacesRequest)
>       returns (GetUnboundWorkspacesResponse) {
>     option (google.api.http) = {
>       get: "/api/v1/workspaces/unbound",
>     };
>     option (grpc.gateway.protoc_gen_swagger.options.openapiv2_operation) = {
>       tags: "Workspaces"
>     };
>   }
> 
>   // Binds the specified workspaces to new auto-created namespaces if they're unbound.
>   rpc BulkAutoCreateWorkspaceNamespaceBindings(
>       BulkAutoCreateWorkspaceNamespaceBindingsRequest)
>       returns (BulkAutoCreateWorkspaceNamespaceBindingsResponse) {
>     option (google.api.http) = {
>       post: "/api/v1/workspaces/auto-namespace-bindings",
>     };
>     option (grpc.gateway.protoc_gen_swagger.options.openapiv2_operation) = {
>       tags: "Workspaces"
>     };
>   }
> 
> // Request for finding all unbound workspaces.
> message GetUnboundWorkspacesRequest {}
> 
> // Response to GetUnboundWorkspacesRequest.
> message GetUnboundWorkspacesResponse {
>   // The list of returned workspace ids.
>   repeated int32 = 1;
>   // Pagination information of the full dataset.
>   Pagination pagination = 2;
> }
> 
> // Request for binding the specified unbound workspaces to new auto-created namespaces.
> message BulkAutoCreateWorkspaceNamespaceBindingsRequest {
>   // The list of returned workspace ids.
>   repeated int32 = 1;
> }
> 
> // Response to BulkAutoCreateWorkspaceNamespaceBindingsRequest.
> message BulkAutoCreateWorkspaceNamespaceBindingsResponse {}
> ```

@hkumar92 What do you think about this API change? We would have one API that would fetch all unboundWorkspaces and one API that would require all those unbound workspace Ids in the request to autocreate namespaces for all

File: master/internal/api_workspace_intg_test.go
Comment (salonig23) [reply master/internal/api_workspace_intg_test.go]: fixed

File: master/internal/api_workspace.go
Comment (salonig23) [reply master/internal/api_workspace.go]: Opted for using Table instead of Model

File: proto/src/determined/api/v1/api.proto
Comment (salonig23) [inline proto/src/determined/api/v1/api.proto:2055]: For some reason this API endpoint is erroring out with: 
```

> {
>   "error": {
>     "code": 3,
>     "reason": "InvalidArgument",
>     "error": "type mismatch, parameter: id, error: strconv.ParseInt: parsing \"default-namespace-bindings\": invalid syntax"
>   }
> }

```
Spent a bunch of time trying to figure out why but no luck. Would appreciate a second pair of eyes @corban-beaird @maxrussell

Comment (maxrussell) [reply proto/src/determined/api/v1/api.proto:2055]: I'm happy to pair! Just LMK on Slack whenever you're ready.

File: webui/react/src/services/api-ts-sdk/api.ts
Comment (keita-determined) [inline webui/react/src/services/api-ts-sdk/api.ts:1]: i wonder why web team is auto assigned for the generated files. 
Is CODEOWNER file wrong?

https://github.com/determined-ai/determined/blob/ca90a63ec5c0c218ef3b8183ac06d5b18dfe3bcf/.github/CODEOWNERS#L49

**Review by keita-determined — APPROVED** [review]

File: webui/react/src/services/api-ts-sdk/api.ts
Comment (salonig23) [reply webui/react/src/services/api-ts-sdk/api.ts:1]: not sure why :/

File: proto/src/determined/api/v1/api.proto
Comment (salonig23) [reply proto/src/determined/api/v1/api.proto]: fixed by changing path

File: master/internal/api_workspace.go
Comment (corban-beaird) [inline master/internal/api_workspace.go:1359]: nit (non-blocking): It might be possible to handle the population of this struct in the SQL; however, by no means do you have to do it that way. It could be to be a trade off between potentially making the SQL query more complex and/or performance. 

If it's easier to approach this using go code, then let's keep this the way it is, but if you do something like an `ARRAY_AGG(cluster_name)` + a `GROUP BY (workspace_id)`, it might help avoid having to do the additional step of transforming the query result into the map.

File: master/internal/api_workspace.go
Comment (corban-beaird) [inline master/internal/api_workspace.go:1389]: nit: It might be worth having the be a `<` check instead, in case we get into a funky case where `len(wkspBinding[wID]) > len(a.m.allRms)`

Comment (corban-beaird) [reply master/internal/api_workspace.go:1389]: Do we have any way of preventing multiple namespace bindings?

Comment (salonig23) [reply master/internal/api_workspace.go:1389]: For a given workspace we should never have more than one binding per cluster. In some cases we might have stale bindings where we might have bindings related to a cluster that is no longer active.

File: master/internal/api_workspace.go
Comment (salonig23) [reply master/internal/api_workspace.go:1359]: I think it is easier to just keep the go code for now, especially given the time crunch 😅

Comment (corban-beaird) [reply master/internal/api_workspace.go:1359]: Absolutely fair!

File: master/internal/api_workspace.go
Comment (salonig23) [reply master/internal/api_workspace.go:1389]: Do you think it is worth having a panic in the case where `len(wkspBinding[wID]) > len(a.m.allRms)` ?

Comment (corban-beaird) [reply master/internal/api_workspace.go:1389]: You made a good point about there being potential stale bindings from an cluster that's no longer active, so I don't know if that's worth adding a panic for, if we can still move forward while one of the clusters is down, we should be solid.

Comment (salonig23) [reply master/internal/api_workspace.go:1389]: I am filtering the wkspBinding map to only have clusters that are currently active (in the sql query from the prev function) so there technically should be no case where `len(wkspBinding[wID]) > len(a.m.allRms)`. So if we do have that something might be going very wrong

Comment (corban-beaird) [reply master/internal/api_workspace.go:1389]: If it's possible for a cluster to become in-active between the API calls, I'd still hold off on the panic; it would be a pretty rough situation to debug

Comment (salonig23) [reply master/internal/api_workspace.go:1389]: The cluster might be active in Kubernetes but to use it in our system again would require a master restart and config change
So I think a panic is appropriate?

Comment (salonig23) [reply master/internal/api_workspace.go]: Didn't add the panic

**Review by corban-beaird — APPROVED** [review]
Looks solid! Great work!

**Pull request merged**
